DOC: Include description of pre-processing transforms (#300)
BrownBiomechanics/Autoscoper: Documentation/transforms.md

and orientation. In DICOM format, the spatial referencing can be retrieved from the DICOM header meta data. In MATLAB,
the `imref3d` function can be constructed from the DICOM meta data to store the intrinsic - spatial relationship.

Transforming between image and world spaces has a visual proportion change to the image volume:

![Image space to World Space](https://)

We can construct the following matrix to convert from image space to world space. Where `P_{x,y,z}` is the pixel spacing in
each respective dimension and `O_{x,y,z}` is the origin for each respective dimension.

```{math}
\begin{bmatrix}
P_x & 0 & 0 & O_x\\
0 & P_y & 0 & O_y\\
0 & 0 & P_z & O_z\\
0 & 0 & 0 & 1
\end{bmatrix}
```

When importing DICOM data with Slicer, a spatial referencing transform is automatically applied to a CT image volume the
header meta data. Note the RAS anatomical orientation and +X,+Y +Z red/green/blue axes indicators.

<!-- Lots of images/diagrams to bring over -->

## Transforms in SlicerAutoscoperM

SlicerAutoscoperM allows tracking of a rigid body in target 'World' space. Often, more than one rigid body

> is tracked, and the relative motion of one rigid body is calculated with respect to a reference.

Author: I believe this is another rigid body.

> is tracked, and the relative motion of one rigid body is calculated with respect to a reference.

Reviewer: @jcfr have these been addressed? Can we commit and make live please?

* `{bone}.tfm` is a non-rigid transform that translates and scales the `{bone}.tif` volume to its spatial location
within the segmented CT-DICOM.
* `{bone}.tra` is the same as the `{bone}_DICOM2AUT.tfm` but in a format which Autoscoper can load.
* `{bone}_t.tfm` is a translation matrix that moves between the world origin and the location of the partial volume
within the segmented CT-DICOM.
* `{bone}_scale.tfm` is a scaling matrix that moves the volume from image space to world space.
* `{bone}_DICOM2AUT.tfm` is a transformation that moves from DICOM space into Autoscoper space.
* `{bone}_PVOL2AUT.tfm` is a transformation that moves from world space into Autoscoper space.

```{note}
The braces {} around the word bone have been omitted below, since they are not supported by mermaid.
```

```{mermaid}
    ---
    title: SlicerAutoscoperM Pre-Processing Transforms
    ---
    flowchart TD
        T[TIFF - Image Space]
        D[Spacially located PV - DICOM Space]
        W[World origin]
        M[Model - AUT Space]

        W -- bone_t.tfm --> D
        T -- bone.tfm --> D
        T -- bone_scale.tfm --> W

        D -- bone_DICOM2AUT.tfm --> M

>         W -- bone_PVOL2AUT.tfm --> M

Reviewer 2: And here is revisited version where we can describe each space by a subgraph:

```
flowchart TD
    subgraph image_space["Image Space"]
      tiff["TIFF"]
    end
    subgraph dicom_space["DICOM Space"]
      transformed_pvol["Spatially located PV"]
    end
    subgraph world_space["World Space"]
      world_origin["World origin"]
      partial_volume["Partial Volume (PVOL)"]
    end
    subgraph aut_space["Autoscoper Space (AUT)"]
      model["Model"]
    end
    world_space -- "{bone}_t.tfm" --> dicom_space
    image_space -- "{bone}.tfm" --> dicom_space
    image_space -- "{bone}_scale.tfm" --> world_space
    image_space -- "{bone}_DICOM2AUT.tfm" --> aut_space
    world_space -- "{bone}_PVOL2AUT.tfm" --> aut_space
```

```mermaid
flowchart TD
    subgraph image_space["Image Space"]
      tiff["TIFF"]
    end
    subgraph dicom_space["DICOM Space"]
      transformed_pvol["Spatially located PV"]
    end
    subgraph world_space["World Space"]
      world_origin["World origin"]
      partial_volume["Partial Volume (PVOL)"]
    end
    subgraph aut_space["Autoscoper Space (AUT)"]
      model["Model"]
    end
    world_space -- "{bone}_t.tfm" --> dicom_space
    image_space -- "{bone}.tfm" --> dicom_space
    image_space -- "{bone}_scale.tfm" --> world_space
    image_space -- "{bone}_DICOM2AUT.tfm" --> aut_space
    world_space -- "{bone}_PVOL2AUT.tfm" --> aut_space
```

# Transforms

## Background

:::{figure-md} Slicer-coordinate-systems
:align: center

![Coordinate systems](https://github.com/Slicer/Slicer/releases/download/docs-resources/coordinate_systems.png)

World coordinate system (left), Anatomical coordinate system (middle), Image coordinate system (right) Image source: [Slicer Coordinate Systems documentation](https://slicer.readthedocs.io/en/latest/user_guide/coordinate_systems.html)
:::

In 3D Slicer, medical image data is processed using the Right-Anterior-Superior (RAS) coordinate
system by default. However, to maintain compatibility with other medical imaging software, Slicer
assumes that files are stored in the Left-Posterior-Superior (LPS) coordinate system unless
otherwise specified.

```{image} https://github.com/BrownBiomechanics/Autoscoper/releases/download/docs-resources/transforms_XYZ_axes.png
:align: center
:alt: XYZ Axes
```

> When reading or writing files, Slicer may need to flip the sign of the first
> two coordinate axes to convert between RAS and LPS. The following transformation
> matrix converts between RAS and LPS coordinates:

Reviewer 2: ```suggestion
When reading or writing files, Slicer automatically converts between RAS and LPS
coordinate systems as needed. This conversion involves flipping the sign of the first
two coordinate axes. The transformation matrix for converting between RAS and LPS
coordinates is shown below:
```

```{image} https://github.com/BrownBiomechanics/Autoscoper/releases/download/docs-resources/transforms_LPS_to_RAS_coords.png
:align: center
:alt: LPS to RAS Coordinate Transformation
```

For more details on the coordinate systems used in 3D Slicer, refer to the [Slicer documentation](https://slicer.readthedocs.io/en/latest/user_guide/coordinate_systems.html).

### Spatial referencing

> Spatial referencing data is used to encode voxel to world unit resolution (also referred to as pixel/voxel spacing), origin,
> and orientation. In DICOM format, the spatial referencing can be retrieved from the DICOM header meta data. In MATLAB,
> the [`imref3d`](https://www.mathworks.com/help/images/ref/imref3d.html) function can be constructed from the DICOM meta data to store the intrinsic spatial relationship.
> 
> Transforming between image space and world space results in proportional changes in the image volume:

Reviewer 2: ```suggestion
Spatial referencing defines the relationship between voxel coordinates in image space
and their corresponding positions in world space. This includes encoding voxel-to-world
unit resolution (also known as pixel/voxel spacing), the origin, and orientation of the
dataset.

In the DICOM format, spatial referencing information is embedded in the DICOM header
metadata. In MATLAB, the [`imref3d`](https://www.mathworks.com/help/images/ref/imref3d.html) function can be constructed
using this metadata to store the intrinsic spatial relationships.

Transforming data between image space and world space involves proportional adjustments
to the image volume, as shown in the matrix below:
```

```{image} https://github.com/BrownBiomechanics/Autoscoper/releases/download/docs-resources/transforms_Image_to_world_space.png
:align: center
:alt: Image space to World Space
```

> The matrix shown above can be used to convert from image space to world space, where `P_{x,y,z}` represents pixel spacing along
> each axis, and `O_{x,y,z}` is the origin for each respective dimension.

Reviewer 2: ```suggestion
In the matrix shown above, `P_{x,y,z}` represents the pixel spacing along each axis, while `O_{x,y,z}` defines the origin for each dimension.
```

> When importing DICOM data into Slicer, a spatial referencing transform is automatically applied to a CT image volume based on the
> header metadata in the DICOM header. Note the RAS anatomical orientation and +X, +Y, +Z (red/green/blue) axes indicators.

Reviewer 2: ```suggestion
When importing DICOM data into 3D Slicer, a spatial referencing transform is automatically applied to the CT image volume.
This transform is derived from the DICOM header metadata and aligns the data to Slicer’s RAS anatomical orientation system,
with the +X, +Y, and +Z directions corresponding to the red, green, and blue axes, respectively.
```

```{image} https://github.com/BrownBiomechanics/Autoscoper/releases/download/docs-resources/transforms_Slicer_DICOM_data.png
:align: center
:alt: DICOM Data Loaded into Slicer
```

> Spatial referencing information of a volume in Slicer can be viewed in the Volume Information drop down of the
> Volumes module.

Reviewer 2: ```suggestion
The spatial referencing information of a volume in Slicer can be inspected in the **Volume Information** section of the **Volumes** module. This provides detailed metadata, including spacing, origin, and orientation.
```

```{image} https://github.com/BrownBiomechanics/Autoscoper/releases/download/docs-resources/transforms_Volume_spatial_info.png
:align: center
:alt: Spatial Referencing Information of a Volume
```

## Transforms in SlicerAutoscoperM

SlicerAutoscoperM enables tracking of rigid bodies in the 'World' space. When tracking multiple
rigid bodies, their relative motions can be computed with respect to a reference body.

> In the AutoscoperM Slicer module, under the Pre-processing tab, rigid bodies of interest can
> be automatically segmented from a CT DICOM volume. Alternatively, previously generated segmentations
> can be loaded. These segmentations are used to generate partial volumes, where the density data
> within the bounds of the segmented model is isolated and saved as a TIFF stack. In Autoscoper, this
> TIFF stack is imported in a specific orientation, referred to as Autoscoper (AUT) space. The data from the
> partial volume is projected onto the target image plane (overlaid with the radiograph for each frame)
> as a Digitally Reconstructed Radiograph (DRR).
> 
> Like CT DICOM volumes, TIFF stacks are initially in image space and require a
> spatial referencing transform to describe their world location, spacing, and
> orientation. In Autoscoper, the voxel resolution of TIFF volumes follows a specified axes.

Reviewer 2: ```suggestion
### Pre-Processing and Volume Generation

In the Pre-processing tab of the SlicerAutoscoperM module, rigid bodies of interest can
be automatically segmented from a CT DICOM volume. Alternatively, previously generated segmentations
can be loaded. These segmentations are used to generate partial volumes, isolating the density data
within the bounds of the segmented model. The partial volume is saved as a TIFF stack, which is later imported into Autoscoper in a specific orientation referred to as Autoscoper (AUT) space.

Within Autoscoper, this TIFF stack is used to generate Digitally Reconstructed Radiographs (DRRs)
by projecting the density data onto the target image plane, where it is overlaid with the radiographs for each frame.

TIFF stacks, like CT DICOM volumes, originate in image space and require a
spatial referencing transform to encode their world location, spacing, and
orientation. In Autoscoper, the voxel resolution of TIFF volumes aligns with predefined axes.
```

```{image} https://github.com/BrownBiomechanics/Autoscoper/releases/download/docs-resources/transforms_Model_and_its_TIFF_pv.png
:align: center
:alt: Segmented Model and Corresponding TIFF Stack
```

> To facilitate post-processing of output transforms from Autoscoper, several folders
> are populated at the time of Partial Volume Generation in SlicerAutoscoperM. The
> following structure outlines the information in the output directory:

Reviewer 2: ```suggestion
### Output Directory Structure

During partial volume generation, SlicerAutoscoperM creates several folders in the output directory. These folders store the models, volumes, transforms, and tracking information required for processing. The structure is as follows:
```

```
Output Directory
│
├── Models
│   └── AUT{bone}.stl
│
├── Tracking
│   └── {bone}.tra
│
├── Transforms
│   ├── {bone}.tfm
│   ├── {bone}_DICOM2AUT.tfm
│   ├── {bone}_PVOL2AUT.tfm
│   ├── {bone}_scale.tfm
│   └── {bone}_t.tfm
│
└── Volumes
    └── {bone}.tif
```

> 

Reviewer 2: ```suggestion

**Folder Descriptions:**
```

* **Models:**
  * `AUT{bone}.stl`: Mesh file of the volume segmentation placed in Autoscoper (AUT) space.
* **Tracking:**
  * `{bone}.tra`: Equivalent to `{bone}_DICOM2AUT.tfm` but formatted for Autoscoper compatibility.
* **Transforms:**
  * `{bone}.tfm`: Non-rigid transform that translates and scales the `{bone}.tif` volume to its spatial location within the segmented CT-DICOM.
  * `{bone}_DICOM2AUT.tfm`: Transformation from DICOM space into Autoscoper (AUT) space.
  * `{bone}_PVOL2AUT.tfm`: Transformation from world space into Autoscoper (AUT) space.
  * `{bone}_scale.tfm`: Scaling matrix that converts the volume from image space to world space.
  * `{bone}_t.tfm`: Translation matrix moving between the world origin and the location of the partial volume within the segmented CT-DICOM.
* **Volumes:**
  * `{bone}.tif`: Non-spatially transformed volumetric data segmented from CT-DICOM.

> 

Reviewer 2: ```suggestion

### Visualizing Transformations

The following diagrams illustrate the transformations between spaces:
```

:::{figure-md} Tfms-to-dicom-space
:align: center

![Transforms of TIFF Stack to DICOM Space](https://github.com/BrownBiomechanics/Autoscoper/releases/download/docs-resources/transforms_Tfms_to_DICOM_space.png)

Transforms to DICOM space: `{bone}.tfm` (pink arrow), `{bone}_t.tfm` (blue arrow)
:::

:::{figure-md} Tfms-to-aut-space
:align: center

![Transforms of PV to AUT Space](https://github.com/BrownBiomechanics/Autoscoper/releases/download/docs-resources/transforms_Tfms_to_AUT_space.png)

Transforms to Autoscoper space: `{bone}_DICOM2AUT.tfm` (orange arrow), `{bone}_t.tfm` (blue arrow) and `{bone}_PVOL2AUT.tfm` (gray arrow)
:::

> 

Reviewer 2: ```suggestion
### Transformation Workflow

The relationships between image, world, and Autoscoper spaces are illustrated below:
```
